python_modules/libraries/dagster-dg/dagster_dg/component.py

```diff
@@ -82,108 +89,97 @@ def __repr__(self) -> str:
 
 def all_components_schema_from_dg_context(dg_context: "DgContext") -> Mapping[str, Any]:
     """Generate a schema for all components in the current environment, or retrieve it from the cache."""
-    schema_raw = None
+    schema = None
     if dg_context.has_cache:
         cache_key = dg_context.get_cache_key("all_components_schema")
-        schema_raw = dg_context.cache.get(cache_key)
-        if schema_raw is None:
-            print("Component schema cache is invalidated or empty. Building cache...")  # noqa: T201
+        schema = dg_context.cache.get(cache_key, dict[str, Any])
 
-    if not schema_raw:
+    if schema is None:
+        if dg_context.has_cache:
+            print("Component schema cache is invalidated or empty. Building cache...")  # noqa: T201
         schema_raw = dg_context.external_components_command(["list", "all-components-schema"])
-    return json.loads(schema_raw)
+        schema = json.loads(schema_raw)
+
+    return schema
 
 
 # ########################
 # ##### HELPERS
 # ########################
 
+MIN_DAGSTER_COMPONENTS_LIST_PLUGINS_VERSION = Version("1.10.12")
```

Comment:
its great that we are considering and testing backcompat - I am pretty worried about this version spaghetti thing become unworkable over time though. Wonder if there is some way for the underlying environment being called to expose its capabilities in a more structured way somehow?

Reply:
Yup that is a problem we will need to solve but it's going to require significant thought to get it right. For now I am just trying to implement the version checks in a consistent way to minimize complexity.